pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/segment/writer/SegmentWriter.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.ingestion.segment.writer;
+
+import java.io.Closeable;
+import java.io.IOException;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.env.PinotConfiguration;
+
+
+/**
+ * An interface to write records into Pinot.
+ * This interface helps abstract out details regarding segment generation and push from the caller.
+ */
+public interface SegmentWriter extends Closeable {
+
+  /**
+   * Initialize the {@link SegmentWriter} with details about the Pinot cluster and table
+   */
+  void init(PinotConfiguration conf)
+    throws IOException;
+
+  /**
+   * Collect a single {@link GenericRow} into a buffer.
+   * This row is not available in Pinot until a <code>flush()</code> is invoked.
```

Comment:
> This row is not available in Pinot until a <code>flush()</code> is invoked.

Does the availability of a row here mean it cannot be queried?

Reply:
That would depend on the impl. For say an OFFLINE table, we'd use something like a FileBasedBuffer, and we'll only get the rows into Pinot for querying after a flush is called.
But for a realtime table, we'd use a MutableSegmentBuffer, in which case as soon as it's collected, we'll be able to query it